src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.element = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndex(index);
+        if (index == size) {
+            link(value);
+        } else {
+            Node<T> node = findNode(index);
+            Node<T> prevNode = node.prev;
+            Node<T> newNode = new Node<>(prevNode, value, node);
+            node.prev = newNode;
+            if (prevNode == null) {
+                head = newNode;
+            } else {
+                prevNode.next = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findNode(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unLink(findNode(index));
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(T value) {
+        Node<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (compareEquality(value, node.element)) {
+                unLink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException("Index of bound. Input index is wrong!");
+        }
+    }
+
+    private void checkAddIndex(int index) {
+        if (!(index >= 0 && index <= size)) {
```

Comment:
better to avoid negating and overcomplicating conditions if we can, check other method as well
```suggestion
        if (index < 0 && index > size) {
```